packages/mcdoc/src/runtime/checker/index.ts

```diff
@@ -0,0 +1,870 @@
+import { ErrorSeverity, type CheckerContext, type FullResourceLocation , type Range } from "@spyglassmc/core";
+import type { Attribute, StructTypePairField , LiteralType, Index, ParallelIndices, KeywordType, UnionType, ListType, EnumType, NumericType, PrimitiveArrayType, StringType, StructType, TupleType } from "../../type/index.js";
+import { NumericRange, McdocType } from "../../type/index.js"
+import { TypeDefSymbolData } from "../../binder/index.js"
+import { arrayToMessage, localeQuote, localize } from "@spyglassmc/locales";
+import { RangeKind, type EnumKind } from "../../node/index.js";
+
+export type NodeEquivalenceChecker = (inferredNode: Exclude<SimplifiedMcdocTypeNoUnion, LiteralType | EnumType>, definition: SimplifiedMcdocTypeNoUnion) => boolean
+
+export type TypeInfoAttacher<T> = (node: T, definition: SimplifiedMcdocType) => void
+
+export type ChildrenGetter<T> = (node: T, simplified: SimplifiedMcdocTypeNoUnion) => RuntimeUnion<T>[]
+
+export type ErrorReporter<T> = (error: ValidationError<T>) => void
+
+export type RuntimeUnion<T> = RuntimeNode<T>[] | RuntimePair<T>
+
+export interface RuntimeNode<T> {
+	originalNode: T;
+	inferredType: Exclude<McdocType, UnionType>;
+}
+export interface RuntimePair<T> {
+	attributes?: Attribute[];
+	key: RuntimeNode<T>;
+	possibleValues: RuntimeNode<T>[];
```

Comment:
I don't think it's needed, I think this might be a remnant from a very early stage of this logic.

Reply:
StructTypePairField also has attributes on itself, so there can be attributes related to the pair itself, not just the key or the value.

These attributes could be added here, currently they are unused since attribute handling is not done yet, but I added them in the isAssignable function so they are at least filled and get a reference in the code.

I am also not sure which attributes actually can exist on a pair field, and I am not sure for which cases it is needed, it's just how the typedefs are currently, and there might be a reason for it.

Reply:
I believe we decided that these two are semantically equivalent, so attributes on a pair should probably be flattened to the types
```rs
struct Foo {
  bar: #[id=item] string,

  #[id=item]
  baz: string,
}
```

Reply:
I think I'll keep it like this for now, if we want to flatten it, I think it makes sense to do it everywhere and change this interface while doing so.